ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,328 @@
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    match expression {
+        Expression::Elem(ident) => parse_elem(ident, expression, lc, symbol_table, i),
+        Expression::NamedTraceAccess(named_trace_access) => {
+            parse_named_trace_access(named_trace_access, expression, lc, symbol_table, i)
+        }
+        Expression::Add(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, lc, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, lc, symbol_table, i)?;
+            Ok(Expression::Add(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::Sub(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, lc, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, lc, symbol_table, i)?;
+            Ok(Expression::Sub(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::Mul(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, lc, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, lc, symbol_table, i)?;
+            Ok(Expression::Mul(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::Exp(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, lc, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, lc, symbol_table, i)?;
+            Ok(Expression::Exp(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::ListFolding(lf_type) => {
+            parse_list_folding(lf_type, expression, symbol_table, i)
+        }
+        _ => Ok(expression.clone()),
+    }
+}
+
+/// Parses an identifier in a list comprehension expression.
+///
+/// # Errors
+/// - Returns an error if the iterable is an identifier and that identifier does not correspond to
+///   a vector.
+/// - Returns an error if the iterable is an identifier but is not of a type in set:
+///   { TraceColumns, IntegrityVariable, PublicInput, RandomValuesBinding }.
+/// - Returns an error if the iterable is a slice and that identifier does not correspond to
+///   a vector.
+/// - Returns an error if the iterable is an identifier but is not of a type in set:
+///   { TraceColumns, IntegrityVariable, PublicInput, RandomValuesBinding }.
+fn parse_elem(
+    ident: &Identifier,
+    expression: &Expression,
+    lc: &ListComprehension,
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    for (member, iterable) in lc.context() {
+        if ident == member {
+            match iterable {
+                Iterable::Identifier(ident) => {
+                    let ident_type = symbol_table.get_type(ident.name())?;
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                i,
+                                CURRENT_ROW,
```

Comment:
Why do we hard-code `CURRENT_ROW` here? Shouldn't the offset come from `trace_columns`?

Reply:
The AST has current row trace columns stored as general element since there's no way to disambiguate between whether it's a trace column or any other symbol table like an intermediate variable. So, if it's stored as a trace column in symbol table, we know that it should have 0 row offset here.